airbyte-workers/src/main/java/io/airbyte/workers/process/ExitCodeWatcher.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.workers.process;
+
+import com.google.common.collect.MoreCollectors;
+import io.fabric8.kubernetes.api.model.ContainerStatus;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.WatcherException;
+import java.util.function.Consumer;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ExitCodeWatcher implements Watcher<Pod> {
+
+  private final Consumer<Integer> onExitCode;
+  private final Consumer<WatcherException> onWatchFailure;
+  private boolean exitCodeRetrieved = false;
+
+  /**
+   *
+   * @param onExitCode callback used to store the exit code
+   * @param onWatchFailure callback that's triggered when the watch fails. should be some failed exit
+   *        code.
+   */
+  public ExitCodeWatcher(final Consumer<Integer> onExitCode, final Consumer<WatcherException> onWatchFailure) {
+    this.onExitCode = onExitCode;
+    this.onWatchFailure = onWatchFailure;
+  }
+
+  @Override
+  public void eventReceived(Action action, Pod resource) {
+    try {
+      if (!exitCodeRetrieved && KubePodProcess.isTerminal(resource)) {
```

Comment:
nit: we should move the `isTerminal` into a `KubePodResourceHelper` class. I spent a few mins scratching my head as I mistook this to be for the KubePodProcess object instead of what is it, which is processing the in-memory object.